### ISSUE TITLE
R + Multithreading

### ISSUE THREAD

**Issue by mcschmitz** [Microsoft/LightGBM#484]
I'm having trouble with multithreading in the R Version, since changing the nthread parameter does not affect the train time at all.

Running the code of the basic walkthrough demo I get:
> system.time(lightgbm(data = dtrain, nrounds = 2000, nthread = 1, objective = "binary", verbose = 0))
       User      System verstrichen 
       3.93        1.89        7.48 

> system.time(lightgbm(data = dtrain, nrounds = 2000, nthread = 4, objective = "binary", verbose = 0))
       User      System verstrichen 
       5.46        3.40        7.11 

> system.time(lightgbm(data = dtrain, nrounds = 2000, nthread = 8, objective = "binary", verbose = 0))
       User      System verstrichen 
       9.04        5.51        8.44 

I am using a Win 7 machine with an Intel i7 4900MQ

Thanks and Grettings,
Manuel

**Comment by mcschmitz** [comment]
On larger datasets an deeper Trees there is only one working thread too. So it seems that this is no issue of communication between the threads.

**Comment by guolinke** [comment]
I believe this is kind of build environment issue (openmp). 
@Laurae2 Do you know how to check it ?

**Comment by Laurae2** [comment]
@guolinke just looking at his logs, it is running multithreaded not singlethread. The user time is larger than what a singlethread would be, it means the overhead is too large (just looking at System time shows it's large). But it's because the dataset is too small.

Demo example should peak already with 2 threads.

@mcschmitz If you believe you have a threading issue, can you provide a reproducible example for a larger dataset (like code for generating a reproducible synthetic dataset with 1M observations and 100 features + training on it so we can check). You can use `rnorm` or `sample` to generate random features, and `set.seed` for reproducibility. Post us the system.time log with different number of threads (like 1, 4, 8).

Example:

```r
set.seed(11111)
random_data <- replicate(100, sample(1:10000, 1000000, rep=TRUE))
labels <- sample(0:1, 1000000, rep=TRUE)

# create lgb.Dataset + training code
```

**Comment by mcschmitz** [comment]
Alright so this works:
```
> require(lightgbm)
Lade nötiges Paket: lightgbm
Lade nötiges Paket: R6
> require(mlbench)
Lade nötiges Paket: mlbench
Warning message:
Paket ‘mlbench’ wurde unter R Version 3.3.3 erstellt 
> set.seed(11111)
> random_data <- mlbench.twonorm(1000000, d = 100)
> lgbData  = lgb.Dataset(data.matrix(random_data$x), label = as.numeric(random_data$classes))
> labels = ifelse(as.numeric(random_data$classes) == 1, 0, 1)
> # Single Thread
> system.time(lightgbm(data = random_data$x, label = labels, num_leaves = 4, learning_rate = 1, nrounds = 200, nthread = 1, objective = "binary", verbose = 0))
       User      System verstrichen 
      98.13        1.33       85.39 
> system.time(lightgbm(data = random_data$x, label = labels, num_leaves = 4, learning_rate = 1, nrounds = 200, nthread = 2, objective = "binary", verbose = 0))
       User      System verstrichen 
      89.81        0.42       57.02 
> # 4 Threads
> system.time(lightgbm(data = random_data$x, label = labels, num_leaves = 4, learning_rate = 1, nrounds = 200, nthread = 4, objective = "binary", verbose = 0))
       User      System verstrichen 
      99.89        0.49       35.38 
```
I took another, but equivalent, example.
I had trouble doing the benchmark on a lgbDataset. This caused a breakdown of my R Session in R Studio, but it worked in the R console. Handing the data and the labels directly to the lightgbm function worked well.
